File: apps/windows/TeslaSync.App/feature-views/SignalSparklinePreview/SignalSparklinePreview.Model.cs
using System.Globalization;
using System.Text.Json;
using TeslaSync.App.Core.Data.State;
using TeslaSync.App.Core.Live;

namespace TeslaSync.App.FeatureViews;

/// <summary>
/// Query shape for the signal-sparkline preview — the native mirror of the web component's hardcoded
/// <c>useSignalHistory(vehicleId, signal, { hours: 1, limit: 30 })</c> call and its
/// <c>numericSeries.length &lt; 2</c> render gate
/// (web/src/features/telemetry/components/SignalSparklinePreview.tsx). Centralised so the source, mapper
/// and tests share one definition.
/// </summary>
public static class SignalSparklinePreviewQuery
{
    /// <summary>Trailing window, in hours, the preview charts (web <c>SPARKLINE_HOURS</c>).</summary>
    public const int Hours = 1;

    /// <summary>Maximum samples requested for the trend (web <c>SPARKLINE_LIMIT</c>).</summary>
    public const int Limit = 30;

    /// <summary>
    /// Minimum numeric samples required to draw a line rather than the em-dash fallback (web
    /// <c>numericSeries.length &lt; 2</c>). A single point has no trend, so it collapses to the empty state.
    /// </summary>
    public const int MinSamples = 2;
}

/// <summary>
/// Surface-local helpers over the canonical <see cref="SignalKind"/> (TeslaSync.App.Core.Live) — the native
/// analogue of the web hook's <c>normalizeSignalKind</c> plus the component's <c>NON_NUMERIC</c> set and the
/// <c>{valueKind}</c> chip text (web/src/api/hooks/useSignals.ts and SignalSparklinePreview.tsx). Kept on the
/// surface (mirroring <c>XRayValueKind</c>) because the Core decoder's equivalents are private; pure and
/// unit-tested without a UI host.
/// </summary>
public static class SignalSparklineKinds
{
    private static readonly Dictionary<string, SignalKind> LongFormByName =
        new(StringComparer.Ordinal)
        {
            ["ValueKindString"] = SignalKind.String,
            ["ValueKindBool"] = SignalKind.Bool,
            ["ValueKindInt32"] = SignalKind.Int,
            ["ValueKindInt64"] = SignalKind.Int,
            ["ValueKindEnum"] = SignalKind.Int,
            ["ValueKindFloat"] = SignalKind.Float,
            ["ValueKindDouble"] = SignalKind.Float,
            ["ValueKindTime"] = SignalKind.Time,
            ["ValueKindUnknown"] = SignalKind.Unknown,
            ["ValueKindCompound"] = SignalKind.Unknown,
            ["ValueKindInvalid"] = SignalKind.Unknown,
        };

    private static readonly Dictionary<string, SignalKind> CompactByName =
        new(StringComparer.OrdinalIgnoreCase)
        {
            ["unknown"] = SignalKind.Unknown,
            ["string"] = SignalKind.String,
            ["bool"] = SignalKind.Bool,
            ["int"] = SignalKind.Int,
            ["float"] = SignalKind.Float,
            ["time"] = SignalKind.Time,
        };

    // Mirrors the iota order in internal/tesla/protomodel/types.go (web normalizeSignalKind's numeric branch).
    private static readonly Dictionary<int, SignalKind> CompactByInt =
        new()
        {
            [0] = SignalKind.Unknown,
            [1] = SignalKind.String,
            [2] = SignalKind.Bool,
            [3] = SignalKind.Int,
            [4] = SignalKind.Int,
            [5] = SignalKind.Float,
            [6] = SignalKind.Float,
            [7] = SignalKind.Int,
            [8] = SignalKind.Unknown,
            [9] = SignalKind.Time,
        };

    /// <summary>
    /// Whether a signal of <paramref name="kind"/> has a meaningful trend line. Mirrors the web
    /// <c>NON_NUMERIC = { 'string', 'unknown', 'time' }</c>: numeric kinds are <see cref="SignalKind.Bool"/>
    /// (charted as 0/1), <see cref="SignalKind.Int"/> and <see cref="SignalKind.Float"/>.
    /// </summary>
    public static bool IsNumeric(SignalKind kind) =>
        kind is SignalKind.Bool or SignalKind.Int or SignalKind.Float;

    /// <summary>
    /// The compact lowercase token shown in the non-numeric chip and its tooltip (web renders the raw
    /// <c>{valueKind}</c>). These are technical type identifiers, rendered verbatim rather than localized —
    /// exactly as the web source does.
    /// </summary>
    public static string Token(SignalKind kind) => kind switch
    {
        SignalKind.String => "string",
        SignalKind.Bool => "bool",
        SignalKind.Int => "int",
        SignalKind.Float => "float",
        SignalKind.Time => "time",
        _ => "unknown",
    };

    /// <summary>
    /// Normalise a raw <c>value_kind</c> / <c>kind</c> JSON value (the long-form <c>"ValueKindFloat"</c>
    /// string, the compact <c>"float"</c> string, or the proto enum integer) to a <see cref="SignalKind"/>.
    /// Total like the web <c>normalizeSignalKind</c>: any unrecognised input becomes
    /// <see cref="SignalKind.Unknown"/>.
    /// </summary>
    public static SignalKind Normalize(JsonElement element)
    {
        switch (element.ValueKind)
        {
            case JsonValueKind.String:
                string text = element.GetString() ?? string.Empty;
                if (CompactByName.TryGetValue(text, out var compact))
                {
                    return compact;
                }

                return LongFormByName.TryGetValue(text, out var longForm) ? longForm : SignalKind.Unknown;

            case JsonValueKind.Number when element.TryGetInt32(out int asInt):
                return CompactByInt.TryGetValue(asInt, out var intForm) ? intForm : SignalKind.Unknown;

            default:
                return SignalKind.Unknown;
        }
    }
}

/// <summary>
/// Pure extraction of the chartable numeric series from a
/// <c>GET /signals/{vehicleID}/{signalName}/history</c> payload — the native port of the web component's
/// <c>envelopesToNumbers</c> (SignalSparklinePreview.tsx): a finite numeric value is taken as-is, a boolean
/// folds to 1 / 0, and every other typed value (string, time, null, compound) is skipped. Tolerant of a
/// partial or schema-drifted payload so a malformed history row never aborts the parse. No UI types —
/// unit-tested without a host.
/// </summary>
public static class SignalSparklineSeries
{
    /// <summary>
    /// Project the <c>data</c> array of a typed history response into the numeric series the sparkline plots,
    /// in wire order. A non-object body or a missing / non-array <c>data</c> field yields an empty series.
    /// </summary>
    public static IReadOnlyList<double> FromHistory(JsonElement response)
    {
        if (response.ValueKind != JsonValueKind.Object
            || !response.TryGetProperty("data", out var data)
            || data.ValueKind != JsonValueKind.Array)
        {
            return Array.Empty<double>();
        }

        var series = new List<double>(data.GetArrayLength());
        foreach (var envelope in data.EnumerateArray())
        {
            if (TryNumericValue(envelope, out double value))
            {
                series.Add(value);
            }
        }

        return series;
    }

    // web envelopesToNumbers operates on already-coerced envelopes: number (finite) and boolean (1/0) are
    // kept, everything else dropped. A numeric kind whose value arrived as a JSON string is coerced first
    // (web coerceValue), so we mirror that single string-to-number path for Int/Float kinds only.
    private static bool TryNumericValue(JsonElement envelope, out double value)
    {
        value = 0;
        if (envelope.ValueKind != JsonValueKind.Object
            || !envelope.TryGetProperty("value", out var raw))
        {
            return false;
        }

        switch (raw.ValueKind)
        {
            case JsonValueKind.Number when raw.TryGetDouble(out double number) && double.IsFinite(number):
                value = number;
                return true;

            case JsonValueKind.True:
                value = 1;
                return true;

            case JsonValueKind.False:
                value = 0;
                return true;

            case JsonValueKind.String:
                return TryParseNumericString(envelope, raw, out value);

            default:
                return false;
        }
    }

    private static bool TryParseNumericString(JsonElement envelope, JsonElement raw, out double value)
    {
        value = 0;
        var kind = envelope.TryGetProperty("kind", out var kindElement)
            ? SignalSparklineKinds.Normalize(kindElement)
            : SignalKind.Unknown;
        if (kind is not (SignalKind.Int or SignalKind.Float))
        {
            return false;
        }

        return double.TryParse(
            raw.GetString(),
            NumberStyles.Float,
            CultureInfo.InvariantCulture,
            out value) && double.IsFinite(value);
    }
}

/// <summary>
/// The lifecycle state the signal-sparkline preview can be in. Every branch maps onto a visible surface —
/// none is hidden. <see cref="Disabled"/> reproduces the web <c>if (!enabled) return null</c> gate (the
/// parent flips the preview on as a category leaf expands), and <see cref="NonNumeric"/> /
/// <see cref="Empty"/> / <see cref="Loaded"/> reproduce the web's chip / em-dash / sparkline branches; the
/// native surface additionally renders an explicit <see cref="Error"/> (retry), <see cref="Stale"/> and
/// <see cref="Offline"/> branch — a strict superset of the web that satisfies the prompt's mandated state set.
/// </summary>
public enum SignalSparklinePreviewState
{
    /// <summary>The preview is gated off by its parent (web <c>!enabled</c>); the surface collapses.</summary>
    Disabled,

    /// <summary>A non-numeric signal (string / unknown / time): the compact <c>(kind)</c> chip.</summary>
    NonNumeric,

    /// <summary>First fetch with nothing cached — the pulsing skeleton sized to the sparkline box.</summary>
    Loading,

    /// <summary>Fewer than two samples in the window — the em-dash fallback (web <c>numericSeries &lt; 2</c>).</summary>
    Empty,

    /// <summary>A fresh series with at least two samples — the trend line.</summary>
    Loaded,

    /// <summary>A cached series past the freshness window — the trend line plus a stale chip.</summary>
    Stale,

    /// <summary>The network failed but a cached series remains — the trend (or em-dash) plus an offline chip.</summary>
    Offline,

    /// <summary>The fetch failed with no cached series — the compact retry affordance.</summary>
    Error,
}

/// <summary>
/// Maps the engine's raw <c>RepositoryResult&lt;JsonElement&gt;</c> emissions to typed
/// <c>RepositoryResult&lt;IReadOnlyList&lt;double&gt;&gt;</c>, preserving the cache-then-network status /
/// freshness while extracting the numeric series. A loaded / cached series with fewer than
/// <see cref="SignalSparklinePreviewQuery.MinSamples"/> points collapses to <see cref="LoadStatus.Empty"/> so
/// the surface shows its em-dash fallback (web parity). The offline branch keeps whatever cached series it has
/// so the surface can still render it under the offline chip. Pure — unit-tested without a network or cache.
/// </summary>
public static class SignalSparklinePreviewResultMapper
{
    /// <summary>Map a raw history emission to a typed numeric-series result.</summary>
    public static RepositoryResult<IReadOnlyList<double>> Map(RepositoryResult<JsonElement> raw)
    {
        ArgumentNullException.ThrowIfNull(raw);

        switch (raw.Status)
        {
            case LoadStatus.Loading:
                return RepositoryResult<IReadOnlyList<double>>.Loading();

            case LoadStatus.Empty:
                return RepositoryResult<IReadOnlyList<double>>.Empty(raw.FetchedAt);

            case LoadStatus.Error:
                return RepositoryResult<IReadOnlyList<double>>.Failure(
                    raw.Error ?? new RepositoryError(RepositoryErrorKind.Unknown, "Unknown error"));
        }

        var series = SignalSparklineSeries.FromHistory(raw.Value);
        var fetchedAt = raw.FetchedAt ?? DateTimeOffset.UtcNow;
        bool enough = series.Count >= SignalSparklinePreviewQuery.MinSamples;

        return raw.Status switch
        {
            LoadStatus.Cached => enough
                ? RepositoryResult<IReadOnlyList<double>>.Cached(series, fetchedAt, raw.IsStale)
                : RepositoryResult<IReadOnlyList<double>>.Empty(fetchedAt),
            LoadStatus.Refreshing => enough
                ? RepositoryResult<IReadOnlyList<double>>.Refreshing(series, fetchedAt, raw.IsStale)
                : RepositoryResult<IReadOnlyList<double>>.Empty(fetchedAt),
            LoadStatus.Offline => RepositoryResult<IReadOnlyList<double>>.OfflineCached(
                series, fetchedAt, raw.Error ?? new RepositoryError(RepositoryErrorKind.Network, "Offline")),
            _ => enough
                ? RepositoryResult<IReadOnlyList<double>>.Loaded(series, fetchedAt)
                : RepositoryResult<IReadOnlyList<double>>.Empty(fetchedAt),
        };
    }
}

/// <summary>
/// Canonical registry metadata for the signal-sparkline preview surface — the native mirror of the web
/// component (web/src/features/telemetry/components/SignalSparklinePreview.tsx). Centralises the stable id and
/// the diagnostics slug so the view and view-model stay free of literal identifiers.
/// </summary>
public static class SignalSparklinePreviewRegistration
{
    /// <summary>Stable surface id.</summary>
    public const string Id = "signal-sparkline-preview";

    /// <summary>Diagnostics surface slug emitted with the <c>view.opened</c> event (P1/S11).</summary>
    public const string Slug = "SignalSparklinePreview";
}

/// <summary>
/// PII-safe diagnostics for the signal-sparkline preview surface (P1/S11 diagnostics contract). Records only
/// the operational <c>view.opened</c> event with the surface slug — never a signal name, value or vehicle id —
/// so a diagnostics line can never leak which vehicle or telemetry field was previewed. Thread-safe.
/// </summary>
public sealed class SignalSparklinePreviewDiagnostics
{
    private readonly Action<string>? _sink;
    private long _viewsOpened;

    /// <summary>Creates the collector over an optional PII-safe diagnostics sink.</summary>
    public SignalSparklinePreviewDiagnostics(Action<string>? sink = null) => _sink = sink;

    /// <summary>Number of times the surface has been opened.</summary>
    public long ViewsOpened => Interlocked.Read(ref _viewsOpened);

    /// <summary>Record that the surface was opened, emitting <c>view.opened slug=SignalSparklinePreview</c>.</summary>
    public void RecordViewOpened()
    {
        Interlocked.Increment(ref _viewsOpened);
        _sink?.Invoke($"view.opened slug={SignalSparklinePreviewRegistration.Slug}");
    }
}

File: apps/android/app/src/main/java/com/teslasync/feature-views/SignalSparklinePreview/SignalSparklinePreviewModel.kt
// Pure, framework-free model + projection for the SignalSparklinePreview feature view — the native analogue
// of everything the web component derives before returning JSX
// (web/src/features/telemetry/components/SignalSparklinePreview.tsx): the `SPARKLINE_LIMIT`/`SPARKLINE_HOURS`
// query window, the `NON_NUMERIC` kind set + `isNumeric` guard, the `envelopesToNumbers` numeric/boolean
// extraction, and the branch precedence (`!enabled` ⇒ nothing, non-numeric ⇒ kind chip, loading ⇒ skeleton,
// `numericSeries.length < 2` ⇒ em-dash, else ⇒ Sparkline). The web component owns a single `useSignalHistory`
// query; this layer projects that one cache-then-network feed onto a [SignalSparklinePreviewState] the thin
// composable renders. No Compose, no Android framework, no HTTP: every declaration here is unit-tested
// off-device in the :app:testReleaseUnitTest gate. Values are the raw SI the backend serves (Phase-42); this
// layer renders them verbatim and performs no unit conversion.
//
// Parity extension — the web component leans on TanStack Query's cache and only renders four visible branches
// (chip / skeleton / em-dash / sparkline). The P3 surface contract additionally mandates an explicit error
// affordance and stale / offline surfaces, so a hard failure with no usable cached series resolves to
// [SignalSparklineMode.Error] (a retry affordance) and a cached series served after a failed refresh / past
// its TTL carries [SparklineFreshness.Offline] / [SparklineFreshness.Stale] — the honest ADR-013 freshness
// contract — while still drawing the last-known line (web parity: TanStack keeps `data` across a refetch).
//
// `MatchingDeclarationName` is suppressed for the co-located supporting types; `InvalidPackageDeclaration` is
// suppressed because this surface's mandated directory (com/teslasync/feature-views/SignalSparklinePreview —
// the P3 prompt's allowed-files path) cannot form a valid Kotlin package (a hyphen and a PascalCase segment
// are illegal in a package identifier), so the package intentionally diverges from the path — exactly as the
// sibling SignalCatalogPanel does.
@file:Suppress("MatchingDeclarationName", "InvalidPackageDeclaration")

package io.teslasync.android.featureviews.signalsparklinepreview

import io.teslasync.android.components.feedback.QueryErrorKind
import io.teslasync.shared.core.data.repo.Resource
import io.teslasync.shared.core.diagnostics.Logger
import io.teslasync.shared.core.net.ApiError
import io.teslasync.shared.core.presentation.signals.SignalEnvelope
import io.teslasync.shared.core.presentation.signals.SignalHistoryRange
import io.teslasync.shared.core.presentation.signals.SignalHistoryResponse
import io.teslasync.shared.core.presentation.signals.SignalKind
import io.teslasync.shared.core.presentation.signals.SignalValue

/**
 * Diagnostics surface slug emitted with the `view.opened` event (P1/S11). Carries no VIN, signal name, or
 * value, so a diagnostics line can never leak the vehicle's live state.
 */
const val SIGNAL_SPARKLINE_PREVIEW_SLUG: String = "SignalSparklinePreview"

/** Em dash shown for the "no samples" empty state — the web `'—'` fallback. */
internal const val EM_DASH: String = "\u2014"

/** The web `SPARKLINE_LIMIT` — the trailing-hour history is capped at 30 samples. */
const val SPARKLINE_LIMIT: Int = 30

/** The web `SPARKLINE_HOURS` — the sparkline covers the last hour. */
const val SPARKLINE_HOURS: Int = 1

/** The history window every preview pulls (web `useSignalHistory(id, signal, { hours: 1, limit: 30 })`). */
val SPARKLINE_RANGE: SignalHistoryRange = SignalHistoryRange(hours = SPARKLINE_HOURS, limit = SPARKLINE_LIMIT)

/** The web `numericSeries.length < 2` guard — a trend line needs at least two finite points. */
internal const val MIN_SPARKLINE_POINTS: Int = 2

/**
 * The non-numeric value kinds (web `NON_NUMERIC` set: `'string' | 'unknown' | 'time'`). A signal of one of
 * these kinds has no meaningful trend line and renders the compact kind chip instead of a sparkline.
 */
val NON_NUMERIC_KINDS: Set<SignalKind> = setOf(SignalKind.String, SignalKind.Unknown, SignalKind.Time)

/** Web `!NON_NUMERIC.has(valueKind)`: bool / int / float render a sparkline; string / time / unknown do not. */
fun isNumericKind(kind: SignalKind): Boolean = kind !in NON_NUMERIC_KINDS

/**
 * Project a typed history series onto the plottable numbers — the verbatim port of the web
 * `envelopesToNumbers`: a finite numeric value passes through, a boolean becomes `1`/`0`, and any other kind
 * (string / time / null) is skipped. Mirrors the web `typeof v === 'number'` / `typeof v === 'boolean'` arms.
 */
fun envelopesToNumbers(data: List<SignalEnvelope>): List<Double> {
    val out = ArrayList<Double>(data.size)
    for (envelope in data) {
        when (val value = envelope.value) {
            is SignalValue.Num -> if (value.value.isFinite()) out.add(value.value)
            is SignalValue.Bool -> out.add(if (value.value) 1.0 else 0.0)
            is SignalValue.Text, SignalValue.Null -> Unit
        }
    }
    return out
}

/**
 * The branch the composable renders — the native superset of the web component's render ternary. [Disabled]
 * mirrors the web `if (!enabled) return null`; [NonNumeric] the kind chip; [Loading] the pulsing skeleton;
 * [Empty] the "no samples" em-dash; [Content] the Sparkline; and [Error] is the P3-mandated retry affordance
 * for a hard failure with no cached series.
 */
enum class SignalSparklineMode { Disabled, NonNumeric, Loading, Error, Content, Empty }

/**
 * The freshness of the cached series behind a [SignalSparklineMode.Content] render. [Fresh] draws the line
 * alone (web parity); [Stale] adds a stale affordance and triggers auto-refresh; [Offline] adds an offline
 * affordance over the last-known line (a cached series served after a failed refresh — ADR-013).
 */
enum class SparklineFreshness { Fresh, Stale, Offline }

/**
 * The immutable state the view-model exposes — the cache-then-network projection of the single
 * `useSignalHistory` feed the web component owns. [series] is the plottable numeric history (kept across a
 * refetch / error so stale / offline still draw the last-known line); [freshness] drives the Content
 * affordance + auto-refresh; [isFetching] guards the auto-refresh against a storm; and [errorKind] classifies
 * a hard failure. [valueKind]/[signal] are carried for the kind chip + the sparkline's accessibility label.
 */
data class SignalSparklinePreviewState(
    val mode: SignalSparklineMode,
    val valueKind: SignalKind,
    val signal: String,
    val series: List<Double>,
    val freshness: SparklineFreshness,
    val isFetching: Boolean,
    val updatedAtMillis: Long?,
    val errorKind: QueryErrorKind?,
)

/**
 * Pure projection from the static props + the single `useSignalHistory` [Resource] to the render-ready
 * [SignalSparklinePreviewState] — the native port of the web component's render branches. Side-effect-free so
 * the gate unit-tests every branch without a device.
 */
object SignalSparklineProjection {
    /**
     * Resolve the render state from [enabled] + [valueKind] + the feed [resource] (web's render ternary).
     * A `null` [resource] means the view-model has not opened the feed (disabled, non-numeric, or no
     * vehicle/signal yet) — the web disabled-query branch, which renders the em-dash empty state.
     */
    fun fromResource(
        enabled: Boolean,
        valueKind: SignalKind,
        signal: String,
        resource: Resource<SignalHistoryResponse>?,
    ): SignalSparklinePreviewState =
        when {
            !enabled -> gated(SignalSparklineMode.Disabled, valueKind, signal)
            !isNumericKind(valueKind) -> gated(SignalSparklineMode.NonNumeric, valueKind, signal)
            resource == null -> gated(SignalSparklineMode.Empty, valueKind, signal)
            else -> fromOpenFeed(valueKind, signal, resource)
        }

    /** Project an opened feed emission onto the render state, keeping the cached series across error/refetch. */
    private fun fromOpenFeed(
        valueKind: SignalKind,
        signal: String,
        resource: Resource<SignalHistoryResponse>,
    ): SignalSparklinePreviewState {
        val series = resource.cached?.let { envelopesToNumbers(it.data) } ?: emptyList()
        val hasSeries = series.size >= MIN_SPARKLINE_POINTS
        val isFetching = resource is Resource.Loading
        val isError = resource is Resource.Error
        val mode =
            when {
                hasSeries -> SignalSparklineMode.Content
                isError -> SignalSparklineMode.Error
                isFetching -> SignalSparklineMode.Loading
                else -> SignalSparklineMode.Empty
            }
        val freshness =
            when {
                mode != SignalSparklineMode.Content -> SparklineFreshness.Fresh
                isError -> SparklineFreshness.Offline
                resource.stale -> SparklineFreshness.Stale
                else -> SparklineFreshness.Fresh
            }
        return SignalSparklinePreviewState(
            mode = mode,
            valueKind = valueKind,
            signal = signal,
            series = series,
            freshness = freshness,
            isFetching = isFetching,
            updatedAtMillis = resource.fetchedAtMillis(),
            errorKind = (resource as? Resource.Error)?.let { queryErrorKindOf(it.error) },
        )
    }

    /** The no-feed states (disabled / non-numeric / no-vehicle) — neutral freshness, empty series. */
    private fun gated(
        mode: SignalSparklineMode,
        valueKind: SignalKind,
        signal: String,
    ): SignalSparklinePreviewState =
        SignalSparklinePreviewState(
            mode = mode,
            valueKind = valueKind,
            signal = signal,
            series = emptyList(),
            freshness = SparklineFreshness.Fresh,
            isFetching = false,
            updatedAtMillis = null,
            errorKind = null,
        )

    /** The freshness stamp of a feed emission (web `dataUpdatedAt`), across loading / success / error. */
    private fun Resource<*>.fetchedAtMillis(): Long? =
        when (this) {
            is Resource.Loading -> fetchedAt
            is Resource.Success -> fetchedAt
            is Resource.Error -> fetchedAt
        }

    /**
     * Classify a feed failure into the recovery bucket the retry affordance keys off — the native analogue of
     * the web `classifyQueryError`. HTTP status drives not-found / unauthorized / server; an open breaker maps
     * to the transient waiting bucket; transport failures map to the generic network bucket.
     */
    fun queryErrorKindOf(error: Throwable?): QueryErrorKind =
        when (error) {
            is ApiError.Http ->
                when {
                    error.status == HTTP_NOT_FOUND -> QueryErrorKind.NotFound
                    error.status == HTTP_UNAUTHORIZED || error.status == HTTP_FORBIDDEN -> QueryErrorKind.Unauthorized
                    error.status >= HTTP_SERVER_ERROR -> QueryErrorKind.ServerError
                    else -> QueryErrorKind.Network
                }
            is ApiError.CircuitOpen -> QueryErrorKind.Waiting
            else -> QueryErrorKind.Network
        }

    private const val HTTP_NOT_FOUND = 404
    private const val HTTP_UNAUTHORIZED = 401
    private const val HTTP_FORBIDDEN = 403
    private const val HTTP_SERVER_ERROR = 500
}

/**
 * Emits the one PII-safe `view.opened` diagnostic with the surface [SIGNAL_SPARKLINE_PREVIEW_SLUG] (P1/S11).
 * Kept free of Compose so it is unit-tested with a recording [Logger]; the view-model calls it from the
 * composable's first-composition effect.
 */
fun recordSignalSparklinePreviewOpened(logger: Logger) {
    logger.info("view.opened", mapOf("surface" to SIGNAL_SPARKLINE_PREVIEW_SLUG))
}
